<a href="https://colab.research.google.com/github/Md-Hasib-Askari/deep_learning_exercises/blob/main/Exercise_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🧠 Exercise 4: Visualize Model Training with TensorBoard


**🎯 Goal:**  
Track training/validation loss and accuracy in real time using TensorBoard while training a neural network on the MNIST dataset.

**Required Libraries:**  
- `tensorflow`
- `tensorboard`


### ✅ Requirements

In [1]:
!pip install tensorflow tensorboard

   ---------------------------------------- 0.0/375.9 MB ? eta -:--:--
   ---------------------------------------- 0.0/375.9 MB ? eta -:--:--
   ---------------------------------------- 0.5/375.9 MB 1.7 MB/s eta 0:03:44
   ---------------------------------------- 1.0/375.9 MB 2.0 MB/s eta 0:03:07
   ---------------------------------------- 1.6/375.9 MB 2.0 MB/s eta 0:03:04
   ---------------------------------------- 1.8/375.9 MB 2.1 MB/s eta 0:02:59
   ---------------------------------------- 2.4/375.9 MB 2.1 MB/s eta 0:02:56
   ---------------------------------------- 2.9/375.9 MB 2.2 MB/s eta 0:02:52
   ---------------------------------------- 3.4/375.9 MB 2.2 MB/s eta 0:02:49
   ---------------------------------------- 3.9/375.9 MB 2.3 MB/s eta 0:02:45
   ---------------------------------------- 4.5/375.9 MB 2.3 MB/s eta 0:02:42
    --------------------------------------- 5.2/375.9 MB 2.3 MB/s eta 0:02:40
    --------------------------------------- 5.8/375.9 MB 2.4 MB/s eta 0:02:37


### **📦 Step 1: Import Libraries**


In [2]:
import tensorflow as tf
from tensorflow.keras import layers, models
import datetime

### **🧹 2. Load & Preprocess the MNIST Dataset**

In [3]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()

# Normalize the data
x_train = x_train.astype("float32") / 255.0
x_test = x_test.astype("float32") / 255.0

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 7s 1us/step


### **🧱 3. Build a Simple ANN**

In [22]:
model = models.Sequential([
    layers.Input(shape=(28, 28)),
    layers.Flatten(),
    layers.Dense(128, activation="tanh"),
    layers.Dense(64, activation="tanh"),
    layers.Dense(10, activation="softmax")
])

### **⚙️ 4. Compile the Model**

In [23]:
model.compile(
    optimizer="adam",
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"]
)

### **📁 5. Setup TensorBoard Logging**

In [24]:
# Define the log directory
log_dir = "logs/fit/" + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
tensorboard_callback = tf.keras.callbacks.TensorBoard(
    log_dir=log_dir, 
    histogram_freq=1
)

### **🚀 6. Train the Model with TensorBoard Callback**

In [25]:
history = model.fit(
    x_train, y_train,
    epochs=5,
    validation_split=0.1,
    callbacks=[tensorboard_callback]
)

Epoch 1/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8778 - loss: 0.4213 - val_accuracy: 0.9635 - val_loss: 0.1280
Epoch 2/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9592 - loss: 0.1339 - val_accuracy: 0.9717 - val_loss: 0.0978
Epoch 3/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9738 - loss: 0.0861 - val_accuracy: 0.9737 - val_loss: 0.0858
Epoch 4/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9818 - loss: 0.0586 - val_accuracy: 0.9727 - val_loss: 0.0886
Epoch 5/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9858 - loss: 0.0452 - val_accuracy: 0.9755 - val_loss: 0.0807


**After 5 epochs**
| Activation Function | Accuracy | Loss   |
|---------------------|----------|--------|
| ReLU                | 98.77%   | 0.0401 |
| Tanh                | 98.58%   | 0.045  |